third_party/move/move-compiler-v2/src/function_checker.rs

```diff
@@ -0,0 +1,227 @@
+                                                (
+                                                    env.get_node_loc(*node_id),
+                                                    "called here".to_owned(),
+                                                )
+                                            })
+                                            .collect();
+                                        let msg = format!(
+                                            "{}function `{}` cannot be called from {}function `{}` \
+                                             because it is private to module `{}`",
+                                            if callee_is_inline { "inline " } else { "" },
+                                            callee_env.get_full_name_with_address(),
+                                            if caller_is_inline { "inline " } else { "" },
+                                            caller_name_str,
+                                            callee_env.module_env.get_full_name_str());
+                                        env.diag_with_labels(
+                                            Severity::Error,
+                                            &callee_env.get_loc(),
+                                            &msg,
+                                            call_details,
+                                        );
+                                    }
+                                    false
+                                },
+                            }
+                        };
+                        if callee_is_accessible {
+                            callees_seen
+                                .entry(callee)
+                                .and_modify(|curr| {
+                                    curr.insert(caller_qfid);
+                                })
+                                .or_insert(BTreeSet::from([caller_qfid]));
+                        }
+                    }
+                }
+            }
+        }
+    }
+    // Check for Unused functions: private (or friendless public(friend)) funs with no callers.
+    for callee in private_funcs {
+        if callees_seen
+            .get(&callee)
+            .map(|s| s.is_empty())
+            .unwrap_or(true)
+        {
+            // We saw no uses of private/friendless function `callee`.
+            let callee_env = env.get_function(callee);
+            let callee_loc = callee_env.get_loc();
+            let callee_is_script = callee_env.module_env.get_name().is_script();
+
+            // Entry functions in a script don't need any uses.
+            // Check others which are private.
+            if !callee_is_script {
+                let is_private = matches!(callee_env.visibility(), Visibility::Private);
+                if warn_about_unused {
+                    let msg = format!(
+                        "Function `{}` is unused: it has no current uses and {} so it can have no future uses.",
```

Comment:
Describe a scenario for me.

Reply:
Say I don't use a fun when compiling now, but I add a use later (similarly, I upload a module now with no use right now, but add a use in later upgrade). The text here makes it sound like it is forbidden to have any future uses, while that may not have been the intent of the text.

Reply:
Adding a reply to the edited conversation: there is no scenario if there is no change (hence, I have no issue with the "no current uses" part, and agree with it). IMHO, the "no future uses" part could be dropped, because it serves to confuse me more than help.